multinet/api/tests/test_workspace.py

```diff
@@ -51,48 +70,281 @@ def test_workspace_rest_create(authenticated_api_client: APIClient):
+
+    writers_names = [writer['username'] for writer in new_writers]
+    for writer in workspace.writers:
+        assert writer.username in writers_names
+
+    readers_names = [reader['username'] for reader in new_readers]
+    for reader in workspace.readers:
+        assert reader.username in readers_names
+
+
+@pytest.mark.django_db
+def test_workspace_rest_put_permissions_maintainer(
+    workspace: Workspace,
+    user: User,
+    user_factory: UserFactory,
+    authenticated_api_client: APIClient,
+):
+    workspace.set_user_permission(user, WorkspacePermission.maintainer)
+    old_owner = user_factory()
+    workspace.set_owner(old_owner)
+    new_owner = user_factory()
+    new_maintainers: List[Dict] = [{'username': user_factory().username} for _ in range(2)]
+    new_writers: List[Dict] = [{'username': user_factory().username} for _ in range(2)]
+    new_readers: List[Dict] = [{'username': user_factory().username} for _ in range(2)]
+    request_data = {
+        'public': True,
+        'owner': {'username': new_owner.username},
+        'maintainers': new_maintainers,
+        'writers': new_writers,
+        'readers': new_readers,
+    }
+    r = authenticated_api_client.put(
+        f'/api/workspaces/{workspace.name}/permissions/', request_data, format='json'
+    )
+    r_json = r.json()
+    assert r.status_code == 200
+    assert r_json['public'] == request_data['public']
+    # maintainers cannot set the owner of a workspace
+    assert workspace.owner == old_owner
+
+    maintainers_names = [maintainer['username'] for maintainer in new_maintainers]
+    for maintainer in workspace.maintainers:
+        assert maintainer.username in maintainers_names
+
+    writers_names = [writer['username'] for writer in new_writers]
+    for writer in workspace.writers:
+        assert writer.username in writers_names
+
+    readers_names = [reader['username'] for reader in new_readers]
+    for reader in workspace.readers:
+        assert reader.username in readers_names
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('permission', [WorkspacePermission.reader, WorkspacePermission.writer])
+def test_workspace_rest_put_permissions_forbidden(
+    workspace: Workspace,
+    user: User,
+    user_factory: UserFactory,
+    authenticated_api_client: APIClient,
+    permission: WorkspacePermission,
+):
+    workspace.set_user_permission(user, permission)
+    new_owner = user_factory()
+    new_maintainers: List[Dict] = [{'username': user_factory().username} for _ in range(2)]
+    public = not workspace.public
+    request_data = {
+        'public': public,
+        'owner': {'username': new_owner.username},
+        'maintainers': new_maintainers,
+        'writers': [],
+        'readers': [],
+    }
+    r = authenticated_api_client.put(
+        f'/api/workspaces/{workspace.name}/permissions/', request_data, format='json'
+    )
+
+    assert r.status_code == 403
+    assert workspace.owner is None
```

Comment:
We should probably create another user (not the user fixture or `new_owner`) and set that as the owner, since a workspace without an owner isn't a state we allow naturally, and testing for that seems confusing.